{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1> Epitope Prediction </h1>\n",
    "\n",
    "This tutorial illustrates the use of epytope to predict HLA-I/II epitopes and how to analyze results. epytope offers a long list of epitope prediction methods and was designed in such a way that extending epytope with your favorite method is easy.\n",
    "\n",
    "This tutorial will entail:\n",
    "- Simple epitope prediction from a list of peptide sequences and protein sequences\n",
    "- Manipulation of the results\n",
    "- Consensus prediction with multiple prediction methods\n",
    "- Integration of a new prediction method\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "<h2> Chapter 1: The basics </h2>\n",
    "<br/>\n",
    "We first start with importing the needed packages."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "from epytope.Core import Allele, Peptide, Protein,generate_peptides_from_proteins\n",
    "from epytope.IO import read_lines, read_fasta\n",
    "from epytope.EpitopePrediction import EpitopePredictorFactory"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lets start of with something simple: Defining peptides, and HLA alleles. You find all basic classes under `epytope.Core`. HLA alleles have to be defined in new nomenclature (see: <a href=\"http://hla.alleles.org/\">http://hla.alleles.org/</a>)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "peptides = [Peptide(\"SYFPEITHI\"),Peptide(\"FIASNGVKL\"), Peptide(\"LLGATCMFV\")]\n",
    "allele = Allele(\"HLA-A*02:01\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Of course we don't have to specify all peptides and alleles by hand. We can also read those from file with `read_lines`. For doing so, the file has to contain the peptide sequences and HLA alleles one per line. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "peptides2 = read_lines(\"./data/peptides.txt\", in_type=Peptide)\n",
    "alleles = read_lines(\"./data/alleles.txt\", in_type=Allele)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Proteins are a bit more complicated. They require a amino acid sequence as well as a transcript and gene ID. If `read_lines` is used generic transcript IDs are generated. Or one can use `read_fasta` to read in protein sequences in FASTA format. But one has to specify at which position of the fasta entries  the transcript ID is (based on the contained |). For example if the fasta file has the following structure\n",
    "\n",
    "\n",
    ">\\>ID_Type|Gene_ID|ID_Type|Trancsript_ID|comment<br/>\n",
    ">AAAAAAAA\n",
    "\n",
    "we would call read_fasta as follows:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[PROTEIN: IPVIIERYKGEKQLPVLDKTKFLVPDHVNMSELVKI (aa-seq)\n",
      "\t\t  NP_852610_part2 (orig transcript)\n",
      "\t\t VARIANTS:\n",
      ", PROTEIN: MKMRFFSSPCGKAAVDPADRCKEVQQIRDQHPSK (aa-seq)\n",
      "\t\t  NP_852610.1 (orig transcript)\n",
      "\t\t VARIANTS:\n",
      "]\n"
     ]
    }
   ],
   "source": [
    "proteins = read_fasta(\"./data/proteins.fasta\", id_position=3, in_type=Protein)\n",
    "print(proteins)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To be able to use proteins for prediction, we first have to generate k-mers of the sequences. For that we use the function `generate_peptides_from_protein`, which generates unique peptide k-mers based on the given protein sequences. These peptides are associated with the proteins from which they origin."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "peptide3 = generate_peptides_from_proteins(proteins, 9)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "epytope has only one entry point to the different prediction methods, namely `EpitopePredictorFactory`. It handles the initialization of the different methods and also collects newly implemented prediction methods if properly implemented. To see which prediction methods epytope supports `EpitopePredictorFactory` can helps here as well:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "netmhc 3.0a,3.4,4.0\n",
      "netmhcpan 2.4,2.8,3.0,4.0\n",
      "netmhcstabpan 1.0\n",
      "netmhcii 2.2\n",
      "netmhciipan 3.0,3.1\n",
      "pickpocket 1.1\n",
      "netctlpan 1.1\n",
      "syfpeithi 1.0\n",
      "bimas 1.0\n",
      "epidemix 1.0\n",
      "hammer 1.0\n",
      "smm 1.0\n",
      "smmpmbec 1.0\n",
      "arb 1.0\n",
      "comblibsidney 1.0\n",
      "tepitopepan 1.0\n",
      "calisimm 1.0\n",
      "mhcnuggets-class-1 2.0,2.3.2\n",
      "mhcnuggets-class-2 2.0,2.3.2\n",
      "mhcflurry 1.2.2,1.4.3\n"
     ]
    }
   ],
   "source": [
    "for name,version in EpitopePredictorFactory.available_methods().items():\n",
    "    print(name, \",\".join(version))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lets select one and do predictions. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>HLA-A*02:01</th>\n",
       "      <th>HLA-B*15:01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Seq</th>\n",
       "      <th>Method</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>(T, V, A, T, L, I, S, I, N)</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>11</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Y, N, T, V, A, T, L, I, S)</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(H, I, S, L, A, N, G, H, D)</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(N, T, V, A, T, L, I, S, I)</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(N, T, H, I, S, L, A, N, G)</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       HLA-A*02:01  HLA-B*15:01\n",
       "Seq                         Method                             \n",
       "(T, V, A, T, L, I, S, I, N) syfpeithi           11            9\n",
       "(Y, N, T, V, A, T, L, I, S) syfpeithi            4            2\n",
       "(H, I, S, L, A, N, G, H, D) syfpeithi            8            6\n",
       "(N, T, V, A, T, L, I, S, I) syfpeithi           18            1\n",
       "(N, T, H, I, S, L, A, N, G) syfpeithi           10            0"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predictor = EpitopePredictorFactory(\"Syfpeithi\")\n",
    "results = predictor.predict(peptides2, alleles=alleles)\n",
    "results.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To initialize particular version of a prediction method, we can use the flag `version=\"\"` when calling the PredictorFactories. If we do not specify any version, epytope will initialize the most recent version that is supported."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>HLA-A*02:01</th>\n",
       "      <th>HLA-B*15:01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Seq</th>\n",
       "      <th>Method</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>(T, V, A, T, L, I, S, I, N)</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>11</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Y, N, T, V, A, T, L, I, S)</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(H, I, S, L, A, N, G, H, D)</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(N, T, V, A, T, L, I, S, I)</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(N, T, H, I, S, L, A, N, G)</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       HLA-A*02:01  HLA-B*15:01\n",
       "Seq                         Method                             \n",
       "(T, V, A, T, L, I, S, I, N) syfpeithi           11            9\n",
       "(Y, N, T, V, A, T, L, I, S) syfpeithi            4            2\n",
       "(H, I, S, L, A, N, G, H, D) syfpeithi            8            6\n",
       "(N, T, V, A, T, L, I, S, I) syfpeithi           18            1\n",
       "(N, T, H, I, S, L, A, N, G) syfpeithi           10            0"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predictor = EpitopePredictorFactory(\"Syfpeithi\", version=\"1.0\")\n",
    "results = predictor.predict(peptides2, alleles=alleles)\n",
    "results.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "External tools like `NetMHC` offer two additional flags when calling `.predict()`, `command=\"/path/to/binary\"` and `options=\"command options\"`. `command=\"\"` specifies the path to an alternative binary that should be used instead of the one that is globally registered. With `options=\"\"` you can specify additional commands that will directly be passed to the command line call without any sanity checks."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>HLA-A*02:01</th>\n",
       "      <th>HLA-B*15:01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Seq</th>\n",
       "      <th>Method</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>(T, V, A, T, L, I, S, I, N)</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>11</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Y, N, T, V, A, T, L, I, S)</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(H, I, S, L, A, N, G, H, D)</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(N, T, V, A, T, L, I, S, I)</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(N, T, H, I, S, L, A, N, G)</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       HLA-A*02:01  HLA-B*15:01\n",
       "Seq                         Method                             \n",
       "(T, V, A, T, L, I, S, I, N) syfpeithi           11            9\n",
       "(Y, N, T, V, A, T, L, I, S) syfpeithi            4            2\n",
       "(H, I, S, L, A, N, G, H, D) syfpeithi            8            6\n",
       "(N, T, V, A, T, L, I, S, I) syfpeithi           18            1\n",
       "(N, T, H, I, S, L, A, N, G) syfpeithi           10            0"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predictor = EpitopePredictorFactory(\"Syfpeithi\", version=\"1.0\")\n",
    "results = predictor.predict(peptides2, alleles=alleles, options=\"-s -b\") \n",
    "results.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2> Chapter 2: Data manipulation and consensus prediction</h2>\n",
    "<br/>\n",
    "The predictor all return a data table like object (DataFrame) storing the peptide and allele objects, as well as the predicted results. Because it is a inherited pandas DataFrame we can do all sorts of nifty thinks. \n",
    "\n",
    "For example exporting the results to csv:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "results.to_csv('./data/test.csv', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Or we could plot the score distribution of a particular HLA allele."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<AxesSubplot:title={'center':'HLA-A*02:01'}>,\n",
       "        <AxesSubplot:title={'center':'HLA-B*15:01'}>]], dtype=object)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAEICAYAAABPgw/pAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAAAbQklEQVR4nO3df5Ac9X3m8feDEEgGjDgLL0QSKE4UEozMj1MAF45rjYktZMqQHJcScQCR3Mlw4IKzXD5MEtvEjitxFT7zy8gcYKAiQ1FnQCosbKscb4FSEVhSJCQhfKdg5VgkUAAjsUDhLPncH9ODW6OZnZ6Z3unt3udV1aWZ7u90PzPb81Hvd7u/rYjAzMzK76CiA5iZWT5c0M3MKsIF3cysIlzQzcwqwgXdzKwiXNDNzCrCBd3MrCJc0BOSdko6p2HeEklrWy1vaCtJz0p6uoNtDkn6haRDmyybK+nLDfMOlXSXpH2SXpD02dSy35K0UtK/SnpF0g8lndBm+/89Wc/eZL2HppZdJWm9pLck3Z31PVl59HOflxSSXpc0IuklSfdJmtHQZlDSkoZ5LffD5DsSyTrr01+2yfu3kl5Opq9LUmr5VyRtkTTa+N0rCxf0/HwYeC/wPkm/266xpLnA7wEBfDI1/0xJfw4cnDz/sKTrksVfBuYBxwMfAT4vaWGybAawCjgBGACeBFaOsf2PA9cCHwXmAu8Drk812QV8Fbir3XuxSaujfR44OSIOp7avHUVtf0bSH0haWm8k6Q8lfTp5mmU/nBERhyfTV8ZotxS4ADgZ+ABwHvDp1PIdwOeB72d4LxOSC3p+LqVWQFcnj9u5BFgH3J1uHxHrgK3AbcBi4FzgptRrvhIRv4iI7cD/ApYkr3syIu6MiFci4t+A/wmcIOk9Y+S9MyK2RcQvgK/U15Ws78GIeBh4OcN7scmp030egIjYR+3g48Rk1sPAKPA14BrgN4DvJG3z3A8vBW6IiOGIeB64gf33+Xsi4lHgtRy2VQgX9BxIehdwIbAimRZLOqTNyy5Jtf+4pIHUsvR4DG8DIeko4NeAzallm4H3t1j/h4EXIuLlJOOHJL2aWv7+JusaGOM/ALN3dLnP1197FLUj5XWp2ZH699/Z/zvQzr9IGpb0HUkzU9v5Y0lPpdo12+dbfX9KyQV9fw9LerU+Ad/K+Lo/BN4CfgQ8Qq275BOtGkv6ELVukwciYgPwz8AfJ8vOpPbr4BXA/ck6rwYOT16+N7WqvcARTdY/G7gVeKePPSLWRsSMVLPDm6yLZuuzSuvLPp/YmGzjJeA44NvJ/AuAQ4A/B24Efg78aYYMLwG/S+279B+p7bsr6gsj4rsR8YFU+2b7/OHpfvSyc0Hf3wURMaM+Af8t4+supVacRyPiLeDBZB6Srkv9wWZ5qv2PIuKl5Pl36+0jYl1EfJXar6BExGMR8TVgJGn77tR2303Dr4eSjqb2JftWRNw3RuaRJuuicX1Wef3a5wFOS7YxjVqX4uOSpkXEQxHxbZKj8qSb5du0EREjEbE+yfAicBXwMUnvbvGSZvv8SFRohMKDiw5QdsnR8NnA6ZL+UzL7XcA0STOTYvy1VPvpwB8BUyS9kMw+FJgh6eSI2AwQETtJ/miUPP+FpN3U/qCzJpl9MrAtte6jqBXzVRHx122ib0te/0BqXS/Wu2jMWul0n28UEf8m6Q7gm8BJwPpk/lCP0eqFudURd32ffzJ5vt/3pwp8hN6ZqZKmpaaDgYuB/0Pt7JJTkum3gGHgoibruIBav/iJqfa/AzxOrV99LPcCfyHpKEm/DfxXan9UJTkq+SHwDxFxbYb3ci/wZ5JOTP4j+Iv6upL1HSxpGjCF2n8+9fdrk0se+/x+JE0BLgPeBJ5t07blfijpDEknSDoo+dvPTcBQROxtsbp7gc9KmiXp14Bl7L/PT022dRBwcLKtKe3ez4QSEZ5qv3HtBM5pmLcEWJtaHg3TV4FngM80Wd/ngfVN5v+A2l/aG+f/EfACcPAYGQ+ldvrWPuBF4LOpZZcmmV6n9qtlfTouWf571H69TK/vs8l69lE7q+DQ1LIvN3m/Xy765+Qpv6lf+3yyLL1v7gN+Cnw8Q8aW+yG1/zx+nqx3N7WCfUzqtZ8CtqWeC/g68EoyfR1QavndTba1pOifUyeTkjdiZmYl5y4XM7OKcEE3M6sIF3Qzs4pwQTczq4jCTkObOXNmzJ07t6jNd+T111/nsMMOKzpGxyZD7g0bNrwUEUePc6Rc5LHPl/VnCs6el7H2+cIK+ty5c1m/fn1Rm+/I0NAQg4ODRcfo2GTILelfxjdNfvLY58v6MwVnz8tY+7y7XMzMKsIF3cysIlzQzcwqwgXdzKwiXNDNzCrCBd3MrCIyF3RJUyT9k6RHmiyTpJsk7ZD0lKTT8o1pNvEkQ7duSk37JF1TdC6bvDo5D/1qYDv73/Gj7lxqd6OfB5xB7W4kZ/SczmwCi4ifURsLvD7G9/PAQ0Vmsskt0xF6coeSTwB3tGhyPnBv1KyjdvedY3PKaFYGHwX+OSJKc6GTVU/WI/RvUhu8vtUNhGcBz6WeDyfzdqcbSVoKLAUYGBhgaGiog6j52/J8qxub7G9gOty8YuU7z+fPOnK8Iu0na760dLaRkZHCP+NulDT3YuCAe7jmvc/n8dn0ul91q6Q/V6A82dsWdEnnAXsiYoOkwVbNmsw74M4ZEXE7cDvAggULouhLaZdc+/1M7ZbNH+WGLb/6qHZ+anCcEu0va760dLaJdLlyJ8qWW9IhwCeBLzQuy3ufz+Oz6XW/6lbZfq5pZcmepcvlLOCTknYC9wNnS/q7hjbDwJzU89nArlwSmk185wIbo3bnebPCtC3oEfGFiJgdEXOp/Vr59xHxJw3NVgGXJGe7nAnsjYjdjesyq6iLaNLdYtZvXY+2KOlygIhYDqwGFgE7gDeo3dHbrPIkvQv4feDTRWcx66igR8QQMJQ8Xp6aH8CVeQYzK4OIeAN4T9E5zMBXipqZVYYLuplZRbigm5lVhAu6mVlFuKCbmVWEC7qZWUW4oJuZVYQLuplZRbigm5lVhAu6mVlFuKCbmVWEC7qZWUW4oJuZVYQLuplZRbigm5lVhAu6mVlFtC3okqZJelLSZknbJF3fpM2gpL2SNiXTF8cnrpmZtZLljkVvAWdHxIikqcBaSY9GxLqGdo9HxHn5RzQzsyzaFvTk9nIjydOpyRTjGcrMzDqXqQ9d0hRJm4A9wJqIeKJJsw8m3TKPSnp/niHNzKy9TDeJjoi3gVMkzQAeknRSRGxNNdkIHJ90yywCHgbmNa5H0lJgKcDAwABDQ0O9pe/RsvmjmdoNTN+/bb9yZ82Xls42MjJS+GfcjTLlTr4TdwAnUfvN9U8j4h8LDWWTVqaCXhcRr0oaAhYCW1Pz96Uer5b0LUkzI+KlhtffDtwOsGDBghgcHOwheu+WXPv9TO2WzR/lhi2/+qh2fmpwnBLtL2u+tHS2oaEhiv6Mu1Gy3DcCP4iICyUdAryr6EA2eWU5y+Xo5CgESdOBc4BnGtocI0nJ49OT9b6ce1qzCUTSu4EPA3cCRMQvI+LVQkPZpJblCP1Y4B5JU6gV6gci4hFJlwNExHLgQuAKSaPAm8Di5I+pZlX2PuBfge9IOhnYAFwdEa/XG+TdzZhHd1SvXXndKlNXWqOyZM9ylstTwKlN5i9PPb4FuCXfaGYT3sHAacBnIuIJSTcC1wJ/WW+QdzdjHt1RvXbldatkXWn7KUt2Xylq1r1hYDh11tf/plbgzQrhgm7WpYh4AXhO0gnJrI8CTxcYySa5js5yMbMDfAZYkZzh8ixwWcF5bBJzQTfrQURsAhYUncMM3OViZlYZLuhmZhXhgm5mVhEu6GZmFeGCbmZWES7oZmYV4YJuZlYRLuhmZhXhgm5mVhEu6GZmFeGCbmZWES7oZmYV4YJuZlYRWe4pOk3Sk5I2S9om6fombSTpJkk7JD0lyYP8m5n1WZbhc98Czo6IEUlTgbWSHo2Idak25wLzkukM4LbkXzMz65O2R+hRM5I8nZpMjTeAPh+4N2m7Dpgh6dh8o5qZ2Vgy3eBC0hRqdzT/TeDW1D0U62YBz6WeDyfzdjesJ9Md0Lc8vzdLrP3Mn3Vkx6/Jevfzgen7t715xcqOtzWe+dLSn2lZ7lTeqKy5zYqWqaBHxNvAKZJmAA9JOikitqaaqNnLmqwn0x3Q+3VX8qzbWTZ/lBu29HZzp/HM12o7ZblTeaOy5jYrWkdnuUTEq8AQsLBh0TAwJ/V8NrCrl2BmZtaZLGe5HJ0cmSNpOnAO8ExDs1XAJcnZLmcCeyNiN2Zm1jdZ+hGOBe5J+tEPAh6IiEckXQ4QEcuB1cAiYAfwBr7zuZlZ37Ut6BHxFHBqk/nLU48DuDLfaGYTn6SdwGvA28BoRCwoNpFNZr39pc/MAD4SES8VHcLMl/6bmVWEj9DNehPAjyQF8O3k1Nx3ZL32Iqs8ztHv9fqGbpX5+oKyZHdBN+vNWRGxS9J7gTWSnomIx+oLs157kVUe5+j36zqPRmW+vqAs2d3lYtaDiNiV/LsHeAg4vdhENpm5oJt1SdJhko6oPwY+Bmwd+1Vm48ddLmbdG6A2FAbUvkvfjYgfFBvJJjMXdLMuRcSzwMlF5zCrc5eLmVlFuKCbmVWEC7qZWUW4oJuZVYQLuplZRbigm5lVhAu6mVlFuKCbmVWEC7qZWUVkuafoHEk/kbRd0jZJVzdpMyhpr6RNyfTF8YlrZmatZLn0fxRYFhEbk4GINkhaExFPN7R7PCLOyz+imZll0fYIPSJ2R8TG5PFrwHZg1ngHMzOzznQ0OJekudRuGP1Ek8UflLQZ2AV8LiK2NXl9pru39OuOKlm3MzC9u0xp45mv1XbKcpeVRmXNbVa0zAVd0uHA94BrImJfw+KNwPERMSJpEfAwMK9xHVnv3tKvO6pk3c6y+aPcsKW3gSnHM1+r7ZTlLiuNyprbrGiZznKRNJVaMV8REQ82Lo+IfRExkjxeDUyVNDPXpGZmNqYsZ7kIuBPYHhHfaNHmmKQdkk5P1vtynkHNzGxsWfoRzgIuBrZI2pTMuw44DiAilgMXAldIGgXeBBZHROQf18zMWmlb0CNiLaA2bW4BbskrlJmZdc5XipqZVYQLuplZRbigm/VA0hRJ/yTpkaKzmLmgm/XmampXT5sVzgXdrEuSZgOfAO4oOosZdHjpv5nt55vA54EjWjXIOtxFVnkMi9CvoTUalXlIh7Jkd0E364Kk84A9EbFB0mCrdlmHu8gqj2ER+jW0RqMyD+lQluzucjHrzlnAJyXtBO4Hzpb0d8VGssnOBd2sCxHxhYiYHRFzgcXA30fEnxQcyyY5F3Qzs4pwH7pZjyJiCBgqOIaZj9DNzKrCBd3MrCJc0M3MKsIF3cysIlzQzcwqwgXdzKwistxTdI6kn0jaLmmbpKubtJGkmyTtkPSUpNPGJ66ZmbWS5Tz0UWBZRGyUdASwQdKaiHg61eZcYF4ynQHclvxrZmZ90vYIPSJ2R8TG5PFr1MZ+ntXQ7Hzg3qhZB8yQdGzuac3MrKWOrhSVNBc4FXiiYdEs4LnU8+Fk3u6G12caSrRfw3tm3c7A9O4ypd28YmXHr1k2v/PtpD+Hsgz52aisuc2KlrmgSzoc+B5wTUTsa1zc5CVxwIyMQ4n2a3jPrNtZNn+UG7aUY5SE9OdQliE/G5U1t1nRMp3lImkqtWK+IiIebNJkGJiTej4b2NV7PDMzyyrLWS4C7gS2R8Q3WjRbBVySnO1yJrA3Ina3aGtmZuMgSz/CWcDFwBZJm5J51wHHAUTEcmA1sAjYAbwBXJZ7UjMzG1Pbgh4Ra2neR55uE8CVeYUyM7PO+UpRM7OKcEE3M6sIF3Qzs4pwQTczqwgXdDOzinBBNzOrCBd0sy5JmibpSUmbk6Glry86k01u5RigxGxiegs4OyJGkuEx1kp6NBlx1KzvXNDNupRcUDeSPJ2aTAcMSmfWLy7oZj2QNAXYAPwmcGtEPNGwPNOQ0Vmlhxbe8vzertbR67DM3coyLHI372n+rCO7TJRdL0M69/M9uaCb9SAi3gZOkTQDeEjSSRGxNbU805DRWaWHFu5mmOludTM8daMswyL3a+jsTvUypHM/35P/KGqWg4h4FRgCFhabxCYzF3SzLkk6OjkyR9J04BzgmUJD2aTmLhez7h0L3JP0ox8EPBARjxScySYxF3SzLkXEU9TusWs2IbjLxcysIlzQzcwqIss9Re+StEfS1hbLByXtlbQpmb6Yf0wzM2snSx/63cAtwL1jtHk8Is7LJZGZmXWl7RF6RDwGvNKHLGZm1oO8znL5oKTNwC7gcxGxrVmjrJdBL5s/2nGAbi7LzbqdgendZSpC+nPo5XLlIpU1t1nR8ijoG4HjkxHnFgEPA/OaNcx6GXS/LpXNup1l80e5YUs5zvBMfw69XK5cpLLmNitaz2e5RMS+iBhJHq8Gpkqa2XMyMzPrSM8FXdIxkpQ8Pj1Z58u9rtfMzDrTth9B0n3AIDBT0jDwJWrjPhMRy4ELgSskjQJvAouTcaLNzKyP2hb0iLiozfJbqJ3WaGZmBfKVomZmFeGCbmZWES7oZmYV4YJuZlYRLuhmZhXhgm5mVhEu6GZmFeGCbmZWES7oZmYV4YJu1iVJcyT9RNJ2SdskXV10JpvcyjEmrNnENAosi4iNko4ANkhaExFPFx3MJicfoZt1KSJ2R8TG5PFrwHZgVrGpbDJzQTfLgaS5wKnAEwVHsUnMXS5mPZJ0OPA94JqI2NewLNNtF7c8vzfTtgamw80rVgKwbH7XkTuWxy0Bs9xasF+3n+xUL7dF7Od7ckE364GkqdSK+YqIeLBxed63XSzqdojd3OKxUZZbC/br9pOd6uW2iP18T+5yMetScqeuO4HtEfGNovOYuaCbde8s4GLgbEmbkmlR0aFs8spyC7q7gPOAPRFxUpPlAm4EFgFvAEvqf/k3q7KIWAuo6BxmdVmO0O8GFo6x/FxgXjItBW7rPZaZmXWqbUGPiMeAV8Zocj5wb9SsA2ZIOjavgGZmlk0efy6fBTyXej6czNvd2DDrKVzdnOZTP5WrE1lP+xqY3l2mIqQ/h/QpbmOZP+vI8Yy0nyyn5zXm7mc+szLLo6A360OMZg3zPoWrX4o6VaxXWXP347Svuiw/28bc/cxnVmZ5nOUyDMxJPZ8N7MphvWZm1oE8Cvoq4BLVnAnsjYgDulvMzGx8ZTlt8T5gEJgpaRj4EjAVICKWA6upnbK4g9ppi5eNV1gzM2utbUGPiIvaLA/gytwSmZlZV3ylqJlZRbigm5lVhAu6mVlFuKCbmVWEC7qZWUW4oJuZVYQLuplZRbigm5lVhAu6mVlFuKCbmVWEC7qZWUW4oJuZVYQLuplZRbigm3VJ0l2S9kjaWnQWM3BBN+vF3cDCokOY1bmgm3UpIh4DXik6h1ld+e58bFYikpYCSwEGBgYYGhpq2m7Z/NFM6xuYnr1tnlrl7sTIyEjb9XTz3rrNtuX5vZnbDkyHm1esZP6sIzveTj/fU6aCLmkhcCMwBbgjIv6mYfkgsBL4eTLrwYj4q64SmVVIRNwO3A6wYMGCGBwcbNpuybXfz7S+ZfNHuWFL/4/Ddn5qsOd1DA0N0er912X9HNK6zdbJtuqfezfb6ud7ynJP0SnArcDvA8PATyWtioinG5o+HhHndZXCzMx6lqUP/XRgR0Q8GxG/BO4Hzh/fWGZm1qksv7vNAp5LPR8GzmjS7oOSNgO7gM9FxLbGBnn3J/ZLUf2WvcqaO4/+0ayy5GnM3c98nZB0HzAIzJQ0DHwpIu4sNpVNZlkKuprMi4bnG4HjI2JE0iLgYWDeAS/KuT+xX4rqt+xV1tx59I9mleVn25i7n/k6EREXFZ3BLC1Ll8swMCf1fDa1o/B3RMS+iBhJHq8GpkqamVtKMzNrK0tB/ykwT9KvSzoEWAysSjeQdIwkJY9PT9b7ct5hzcystba/j0fEqKSrgB9SO23xrojYJunyZPly4ELgCkmjwJvA4oho7JYxM7NxlKljOOlGWd0wb3nq8S3ALflGMzOzTvjSfzOzinBBNzOrCBd0M7OKcEE3M6sIF3Qzs4pwQTczqwgXdDOzinBBNzOrCBd0M7OKcEE3M6sIF3Qzs4pwQTczqwgXdDOzinBBNzOrCBd0M7OKcEE3M6sIF3Qzs4rIVNAlLZT0M0k7JF3bZLkk3ZQsf0rSaflHNZt42n03zPqpbUGXNAW4FTgXOBG4SNKJDc3OBeYl01Lgtpxzmk04Gb8bZn2T5Qj9dGBHRDwbEb8E7gfOb2hzPnBv1KwDZkg6NuesZhNNlu+GWd8oIsZuIF0ILIyI/5I8vxg4IyKuSrV5BPibiFibPP8x8D8iYn3DupZSO4IHOAH4WV5vZJzNBF4qOkQXJkPu4yPi6PEM00rG70be+3xZf6bg7Hlpuc8fnOHFajKv8X+BLG2IiNuB2zNsc0KRtD4iFhSdo1POPe7a7vd57/Ml+mwO4OzjL0uXyzAwJ/V8NrCrizZmVeP93iaULAX9p8A8Sb8u6RBgMbCqoc0q4JLkbJczgb0RsTvnrGYTTZbvhlnftO1yiYhRSVcBPwSmAHdFxDZJlyfLlwOrgUXADuAN4LLxi1yI0nUTJZx7HLX6bozzZkvx2bTg7OOs7R9FzcysHHylqJlZRbigm5lVhAt6A0l3SdojaWtq3n+QtEbS/03+ParIjI0kzZH0E0nbJW2TdHUyf6LnnibpSUmbk9zXJ/MndO4ilHmIAUk7JW2RtEnS+vavKFYZa0CdC/qB7gYWNsy7FvhxRMwDfpw8n0hGgWUR8TvAmcCVySXoEz33W8DZEXEycAqwMDlLaqLn7quKDDHwkYg4pQznclPOGgC4oB8gIh4DXmmYfT5wT/L4HuCCfmZqJyJ2R8TG5PFrwHZgFhM/d0TESPJ0ajIFEzx3ATzEQB+VsQbUuaBnM1A/rz75970F52lJ0lzgVOAJSpBb0hRJm4A9wJqIKEXuPpsFPJd6PpzMK4sAfiRpQzIUQhmVYp/Mcum/lYSkw4HvAddExD6p2ZXpE0tEvA2cImkG8JCkkwqONBFlGlpjAjsrInZJei+wRtIzyVGw5cxH6Nm8WB89Mvl3T8F5DiBpKrViviIiHkxmT/jcdRHxKjBEre+yNLn7pNRDDETEruTfPcBD1LqQyqYU+6QLejargEuTx5cCKwvMcgDVDsXvBLZHxDdSiyZ67qOTI3MkTQfOAZ5hgucuQGmHGJB0mKQj6o+BjwFbx37VhFSKfdJXijaQdB8wSG24zBeBLwEPAw8AxwH/D/jPEdH4R5PCSPoQ8DiwBfj3ZPZ11PrRJ3LuD1D7A9MUagcXD0TEX0l6DxM4dxEkLQK+ya+GGPjrYhNlI+l91I7KodbF+92Jnr2MNaDOBd3MrCLc5WJmVhEu6GZmFeGCbmZWES7oZmYV4YJuZlYRLuhmZhXhgm5mVhH/HxNZq4sPUDZJAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "results.hist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To combine prediction results we can use `merge_results` from `epytope.Core`. In addition to the result object we want to merge, also have to specify the type of these objects (here `EpitopePredictionResult`). The function will return a merged results object of the same type."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>HLA-A*02:01</th>\n",
       "      <th>HLA-B*15:01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Seq</th>\n",
       "      <th>Method</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">(F, I, A, S, N, G, V, K, L)</th>\n",
       "      <th>bimas</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>smm</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">(L, L, G, A, T, C, M, F, V)</th>\n",
       "      <th>bimas</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>smm</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">(S, Y, F, P, E, I, T, H, I)</th>\n",
       "      <th>bimas</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>smm</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       HLA-A*02:01  HLA-B*15:01\n",
       "Seq                         Method                             \n",
       "(F, I, A, S, N, G, V, K, L) bimas              NaN          NaN\n",
       "                            smm                NaN          NaN\n",
       "                            syfpeithi          NaN          NaN\n",
       "(L, L, G, A, T, C, M, F, V) bimas              NaN          NaN\n",
       "                            smm                NaN          NaN\n",
       "                            syfpeithi          NaN          NaN\n",
       "(S, Y, F, P, E, I, T, H, I) bimas              NaN          NaN\n",
       "                            smm                NaN          NaN\n",
       "                            syfpeithi          NaN          NaN"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = [EpitopePredictorFactory(m).predict(peptides,alleles=alleles) \n",
    "                    for m in [\"Syfpeithi\",\"BIMAS\",\"SMM\"]]\n",
    "df = results[0].merge_results(results[1:])\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "We also can filter the predicted epitopes based on their prediction values with the function `filter_result` from `epytope.Core`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>HLA-A*02:01</th>\n",
       "      <th>HLA-B*15:01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Seq</th>\n",
       "      <th>Method</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty EpitopePredictionResult\n",
       "Columns: [HLA-A*02:01, HLA-B*15:01]\n",
       "Index: []"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#you can either use pre-defined operators from `operator`\n",
    "from operator import ge\n",
    "#or define you own comparator function like this\n",
    "comparator = lambda a,b: a > b\n",
    "\n",
    "df.filter_result([(\"syfpeithi\",comparator,0.0)])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With that one can combine several prediction tools to form a consensus prediction method.<br/><br/>\n",
    "<h2> Chapter 3: Implementation of a new epitope prediction method </h2>\n",
    "\n",
    "epytope possesses a potent plugin system allowing the user to extend its capability quite easily. To include a new epitope prediction method one simply has to inherit from `epytope.Core.AEpitopePrediction` and implement its interface. For methods calling an external prediction tool additionally have to inherit from `epytope.Core.AExternal`. \n",
    "\n",
    "If you want to be very specific and fully integrate your method in all of epytope's capabilities please use one of the three major interfaces `APSSMEpitopePrediction`, or `AExternalEpitopePrediction` from `epytope.EpitopePrediction`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "from epytope.EpitopePrediction import APSSMEpitopePrediction\n",
    "from epytope.Core import EpitopePredictionResult\n",
    "import random\n",
    "import pandas\n",
    "\n",
    "class RandomEpitopePrediction(APSSMEpitopePrediction):\n",
    "    __alleles = [\"A*02:01\"]\n",
    "    __supported_length = [9]\n",
    "    __name = \"random\"\n",
    "    __version= \"1.0\"\n",
    "    \n",
    "    #the interface defines three class properties\n",
    "    @property\n",
    "    def name(self):\n",
    "        #returns the name of the predictor\n",
    "        return self.__name\n",
    "    \n",
    "    @property\n",
    "    def supportedAlleles(self):\n",
    "        #returns the supported alleles as strings (without the HLA prefix)\n",
    "        return self.__alleles\n",
    "    \n",
    "    @property\n",
    "    def supportedLength(self):\n",
    "        #returns the supported epitope lengths as iterable\n",
    "        return self.__supported_length\n",
    "    \n",
    "    @property\n",
    "    def version(self):\n",
    "        #returns the version of the predictor\n",
    "        return self.__version\n",
    "    \n",
    "    #the interface defines a function converting epytope's HLA allele presentation\n",
    "    #into an internal presentation used by different methods.\n",
    "    #for this predictor we won't need it but still have to provide it!\n",
    "    #the function consumes a list of alleles and converts them into the internally used presentation\n",
    "    def convert_alleles(self, alleles):\n",
    "        #we just use the identity function\n",
    "        return alleles\n",
    "    \n",
    "    #additionally the interface defines a function `predict` \n",
    "    #that consumes a list of peptides or a single peptide and optionally a list \n",
    "    #of allele objects\n",
    "    #\n",
    "    #this method implements the complete prediction routine\n",
    "    def predict(self, peptides, alleles=None):\n",
    "        \n",
    "        #test whether one peptide or a list\n",
    "        if isinstance(peptides, str):\n",
    "            peptides = list(peptides)\n",
    "        \n",
    "        #if no alleles are specified do predictions for all supported alleles\n",
    "        if alleles is None:\n",
    "            alleles = self.supportedAlleles\n",
    "        else:\n",
    "            #filter for supported alleles\n",
    "            alleles = filter(lambda a: a.name in self.supportedAlleles, alleles) \n",
    "        \n",
    "        result = {}\n",
    "        #now predict binding/non-binding for each peptide at random\n",
    "        for a in alleles:\n",
    "            result[a] = {}\n",
    "            for p in peptides:\n",
    "                if random.random() >= 0.5:\n",
    "                    result[a][p] = 1.0\n",
    "                else:\n",
    "                    result[a][p] = 0.0\n",
    "        \n",
    "        #create EpitopePredictionResult object. This is a multi-indexed DataFrame \n",
    "        #with Peptide and Method as multi-index and alleles as columns\n",
    "        df_result = EpitopePredictionResult.from_dict(result)\n",
    "        df_result.index = pandas.MultiIndex.from_tuples([tuple((i,self.name)) for i in df_result.index],\n",
    "                                                        names=['Seq','Method'])\n",
    "        return df_result\n",
    "            \n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now lets use our new predictor."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>A*02:01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Seq</th>\n",
       "      <th>Method</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>(F, I, A, S, N, G, V, K, L)</th>\n",
       "      <th>random</th>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(L, L, G, A, T, C, M, F, V)</th>\n",
       "      <th>random</th>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(S, Y, F, P, E, I, T, H, I)</th>\n",
       "      <th>random</th>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    A*02:01\n",
       "Seq                         Method         \n",
       "(F, I, A, S, N, G, V, K, L) random      1.0\n",
       "(L, L, G, A, T, C, M, F, V) random      1.0\n",
       "(S, Y, F, P, E, I, T, H, I) random      1.0"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EpitopePredictorFactory(\"random\").predict(peptides)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The predictor is now fully integrated and can be used in any context defined by epytope."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
